### PR TITLE
Add batching to `RendererCanvasRenderRD`

### LINKED ISSUE #71084: Vulkan: Editor constantly uses 1-2% CPU while doing nothing on empty project due to Vulkan lacking 2D batching
### Godot version

v4.0.beta10.mono.official [d0398f62f]

### System information

MacOS Catilina 10.15.6, MacBook Air 

### Issue description

I just downloaded 4.0 beta 10 and started a new project, and simply sitting and staring at the empty project screen causes my mac's fan to spin up and stay spinning on high. The editor isn't doing anything, but it apparently just can't sit still? The CPU usage is always at at least 1% or 2%, and it apparently wakes the processor at least 150 times per second. 

I would expect simply staring at the editor with *nothing* on screen would leave the CPU alone. If I swipe away to another screen or minimize the editor, I would expect the CPU usage to fall as well, but the editor still uses 1-2% constantly. There's no way I can use Godot if it is going to eat my laptop's battery just doing literally nothing...

image

### Steps to reproduce

Open a new project and just sit and wait. Going to the AssetLib tab seems to speed up the process, but it's not required.

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by stuartcarnie**
This PR adds batching support for `RendererCanvasRenderRD`. It borrows much of the implementation from the OpenGL compatibility renderer, with additional optimisations to reduce branching and extra work in the hot loops.

Issue #85871 was a particularly interesting case, as zoomed to about 9% on my MacBook Pro M1:

| area | before | after |
| :--- | :--- | :--- |
| draw calls (for tile map) | > 573440 | 36 |
| FPS | ~10 | > 130 |

- *Production edit: This closes https://github.com/godotengine/godot/issues/71084.*

### CODE REVIEW

**Comment by akien-mga** [comment]
GCC warning to fix:
```
servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp: In member function 'void RendererCanvasRenderRD::_allocate_instance_buffer()':
Error: servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp:3903:47: error: comparison of integer expressions of different signedness: 'int' and 'unsigned int' [-Werror=sign-compare]
 3903 |  if (int(state.current_instance_buffer_index) < state.canvas_instance_data_buffers[state.current_data_buffer_index].instance_buffers.size()) {
      |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Review by Calinou — COMMENTED** [review]
Tested locally, it works as expected.

**Testing project:** [test_2d_batching_1x1_sprite.zip](https://github.com/user-attachments/files/15595697/test_2d_batching_1x1_sprite.zip)
This project renders 10,000 Sprite2D nodes that are 1x1 each at a fixed position. This creates a stress point for the CPU while not using the GPU much (as each sprite is tiny).

## Benchmark

Using a Linux optimized editor build (`optimize=speed lto=full`).

<details open>

<summary>PC specifications</summary>

- **CPU:** Intel Core i9-13900K
- **GPU:** NVIDIA GeForce RTX 4090
- **RAM:** 64 GB (2×32 GB DDR5-5800 C30)
- **SSD:** Solidigm P44 Pro 2 TB
- **OS:** Linux (Fedora 40)
</details>

Results are in 3840x2160 fullscreen, but results in 1152x648 are pretty much identical since this is a CPU-bottlenecked scenario.

Before | After *(this PR, old implementation)* | After *(this PR, new implementation)* | Compatibility[^2]
-|-|-|-
935 FPS (1.07 mspf) | 2145 FPS (0.47 mspf) | 1632 FPS (0.62 mspf) | 1563 FPS (0.64 mspf)

Software rendering with llvmpipe (1152x648):

Before | After *(this PR, old implementation)* | After *(this PR, new implementation)* | Compatibility[^2]
-|-|-|-
27 FPS (37.04 mspf) | 183 FPS (5.46 mspf) | 185 FPS (5.41 mspf) | N/A

[^1]: Results between Forward+ and Mobile are largely identical, so they are consolidated here.
[^2]: Not affected by this PR, but this is present for reference.

**Comment by stuartcarnie** [comment]
> GCC warning to fix:
> 
> ```
> servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp: In member function 'void RendererCanvasRenderRD::_allocate_instance_buffer()':
> Error: servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp:3903:47: error: comparison of integer expressions of different signedness: 'int' and 'unsigned int' [-Werror=sign-compare]
>  3903 |  if (int(state.current_instance_buffer_index) < state.canvas_instance_data_buffers[state.current_data_buffer_index].instance_buffers.size()) {
>       |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```

Fixed!

Strange, as that code is a duplicate of the gles3, so they must have that warning disabled?:

https://github.com/godotengine/godot/blob/96a386f3c424af96d950ee5098b4b0e4907c9508/drivers/gles3/rasterizer_canvas_gles3.cpp#L2589

**Comment by AThousandShips** [comment]
> Strange, as that code is a duplicate of the gles3

It isn't fully, the forward code uses a `LocalVector` but gles uses `Vector`

File: servers/rendering/renderer_rd/renderer_canvas_render_rd.h
Comment (alula) [inline servers/rendering/renderer_rd/renderer_canvas_render_rd.h:49]: location 4 is already occupied by `uv_attrib`, I'm getting custom shader compilation errors with this PR.

Comment (stuartcarnie) [reply servers/rendering/renderer_rd/renderer_canvas_render_rd.h:49]: @alula `INSTANCE_DATA_UNIFORM_SET ` is a UBO for a descriptor set 4. `uv_attrib` is a vertex attribute at location 4. There is a vertex stage output variable at location 4:


```
layout(location = 4) out flat uint instance_index_interp;
```

Can you share the specific compilation errors

Comment (alula) [reply servers/rendering/renderer_rd/renderer_canvas_render_rd.h:49]: Oh well I just did some debugging while making a MRP instead of doing it on large project and I realized I were confidently wrong :)

Well, turns out adding any `varying` to the shader triggers following compilation error:

```
ERROR: Failed parse:
WARNING: 0:270: '' : all default precisions are highp; use precision statements to quiet warning, e.g.:
         "precision mediump int; precision highp float;" 
ERROR: 0:270: 'location' : overlapping use of location 4
ERROR: 1 compilation errors.  No code generated.



   at: _compile_variant (servers/rendering/renderer_rd/shader_rd.cpp:294)
```

Quick look, let's see where are `varying` indices are being assigned by shader compiler...

https://github.com/godotengine/godot/blob/e96ad5af98547df71b50c4c4695ac348638113e0/servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp#L2633

After changing that to `5` everything seems to work as expected.

Comment (alula) [reply servers/rendering/renderer_rd/renderer_canvas_render_rd.h:49]:
```glsl
shader_type canvas_item;

varying flat vec4 test1;

void vertex() {
	// Called for every vertex the material is visible on.
}

void fragment() {
	// Called for every pixel the material is visible on.
}
```

Comment (stuartcarnie) [reply servers/rendering/renderer_rd/renderer_canvas_render_rd.h:49]: @alula great find, thanks! Fixed in latest push

File: doc/classes/ProjectSettings.xml
Comment (clayjohn) [inline doc/classes/ProjectSettings.xml:2314]: The equivalent for this in the compatibility renderer is `rendering/gl_compatibility/item_buffer_size`. I wish I had had the foresight to make the setting something backend-agnostic like the name you use here. 

After we merge this PR, we should deprecate `rendering/gl_compatibility/item_buffer_size` and use this setting for both the RD backend and the compatibility backend. I think it should be possible to stop writing `rendering/gl_compatibility/item_buffer_size` and to just read it if present, then delete it.

**Comment by stuartcarnie** [comment]
> [!IMPORTANT]
>
> I've pushed an alternative version to this branch that is more efficient and reduces branching in the CPU side.
>
> Further, it uses `bitfieldExtract` to reduce branching on the GPU side.

File: servers/rendering/renderer_rd/shaders/canvas.glsl
Comment (stuartcarnie) [inline servers/rendering/renderer_rd/shaders/canvas.glsl:138]: Moved to here, as `USE_ATTRIBUTES` the FLAGS_INSTANCING_MASK is only used when using vertex attributes

File: servers/rendering/renderer_rd/shaders/canvas.glsl
Comment (stuartcarnie) [inline servers/rendering/renderer_rd/shaders/canvas.glsl:546]: As tested by @clayjohn this uses fewer instructions than shift + and operators

File: servers/rendering/renderer_rd/shaders/canvas.glsl
Comment (stuartcarnie) [inline servers/rendering/renderer_rd/shaders/canvas.glsl:348]: Using `bitfieldExtract` we avoid a branch

File: servers/rendering/renderer_rd/shaders/canvas.glsl
Comment (stuartcarnie) [inline servers/rendering/renderer_rd/shaders/canvas.glsl:175]: Using `bitfieldExtract` we remove two branches

File: servers/rendering/renderer_rd/storage_rd/texture_storage.cpp
Comment (stuartcarnie) [inline servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:646]: This is a copy of `canvas_texture_get_uniform_set`, but does not create a uniform set, and also provides the cached RIDs and state, to be used by the batch rendering. 

The intent is that a future iteration of the canvas batching could combine multiple textures, samplers, etc into an array, further reducing draw calls. Rendering large blocks of text would benefit from this optimisation, as they switch between textures for the same `TYPE_RECT`, resulting in many individual draw calls.

File: servers/rendering/renderer_rd/renderer_canvas_render_rd.h
Comment (stuartcarnie) [inline servers/rendering/renderer_rd/renderer_canvas_render_rd.h:373]: This key determines when the texture state for the rendering pipeline must change, and reduces the number of branches for testing individual state changes when batching up draw calls.

**Review by stuartcarnie — COMMENTED** [review]
I've added some inline comments

**Comment by Calinou** [comment]
I've updated https://github.com/godotengine/godot/pull/92797#pullrequestreview-2100153431 with new results, but the new implementation appears to be slower than the previous one.

**Comment by stuartcarnie** [comment]
Can you tell if the difference GPU or CPU related?

I can port the `canvas.glsl` changes to the original branch for you to compare.

**Comment by stuartcarnie** [comment]
@Calinou I looked at the shader code, and all the changes are behind `#ifdef`s that wouldn't be used by your benchmark, so it must be the CPU-side changes that have slowed your benchmark down. I'll revert that side back and push up to this branch.

**Comment by Calinou** [comment]
> @Calinou I looked at the shader code, and all the changes are behind `#ifdef`s that wouldn't be used by your benchmark, so it must be the CPU-side changes that have slowed your benchmark down. I'll revert that side back and push up to this branch.

I've updated my review accordingly. It seems better now, but still not at the level of the old implementation.

**Comment by KeyboardDanni** [comment]
It might be useful to keep `rendering/gl_compatibility/item_buffer_size` in some form, since the ideal batch size might differ between OpenGL and Vulkan or other RenderingDevice APIs. Maybe we could roll it into `rendering/2d/batching/item_buffer_size` and add a new platform override for `compatibility`?

**Comment by passivestar** [comment]
Hey, I tried doing a little stress test, batching doesn't seem to be working for text?

I tried with a single label with shadow, 50 paragraphs (23412 characters):

| Master (Vulkan) | This PR | Unity URP (for reference) |
|--------|--------|--------|
| **30** fps | **40** fps | **246** fps |
| **60495** total draw calls | **40332** total draw calls | **5** total draw calls (**1** for the text) |
|  | |  |

**Comment by stuartcarnie** [comment]
@passivestar could you share your project?

The issue is that batching groups sequential draw calls with the same state (texture, modulation, etc). Likely what is happening here is that words are separate draw calls, interleaved with drawing the shadow. Ideally the text system would draw all the shadows first and then draw the text over the top. Another issue with text, is switching textures to draw other fonts, such as bold, which I noticed when rendering the help. 


I believe I can implement an improved version of batching, which can handle multiple textures and other state in a single draw call. I was going to do that as a follow up to this one, as this PR does make a huge difference to TileMap and rendering the same sprite (or atlas).

**Comment by passivestar** [comment]
> Likely what is happening here is that words are separate draw calls, interleaved with drawing the shadow. Ideally the text system would draw all the shadows first and then draw the text over the top

OMG you're right! Disabling the shadow reduces the draw calls to **2** and brings fps to **900**!

sc<br>

> @passivestar could you share your project?

Attached above 👍

**Review by clayjohn — APPROVED** [review]
The new batching implementation looks perfect. I tested on a few example projects locally and did not find any visual regressions and performance was improved as expected (as well as the monitor reporting a massive decrease in draw calls).

This should be ready to merge once you remove the fallback to the old method. 

I'd love to include this in dev3!

**Comment by stuartcarnie** [comment]
Rebased so the build can run; I'll remove the fallback code and squash the commits next.

**Review by clayjohn — APPROVED** [review]
Tested after the rebase and it looks and runs great!

**Comment by akien-mga** [comment]
Tested briefly on Linux (AMD Radeon RX 7600M XT), with unoptimized `dev_build`.

Simple 2D scene, editor FPS with Update Continuously and VSync off:
- FPS `master`: 460-500
- FPS with this PR: 600-650

MRP from #85871, zoomed out to around 7%, with grid enabled, and using `forward_plus`:
- FPS `master`: 3-4
- FPS with this PR: 12-13
Same project with 32% zoom:
- FPS `master`: ~40
- FPS with this PR: ~70

File: doc/classes/ProjectSettings.xml
Comment (stuartcarnie) [reply doc/classes/ProjectSettings.xml]: Fixed

**Comment by akien-mga** [comment]
Amazing work once again 👏 🎉

**Comment by mihe** [comment]
I'm seeing a regression in font subpixel antialiasing that seems to have originated from this pull request, where I get these weird black artifacts around a lot of the text in the editor.

I opened an issue for it here: #97195

**Comment by darksylinc** [comment]
I'm afraid this PR broke mobile support because it uses slots beyond the minimum guaranteed maxBoundDescriptorSets = 4.

I create a ticket for that issue here: #97341

**Comment by LeeWannacott** [comment]
@stuartcarnie @clayjohn Awesome work!, any chance this batching could be extended to the other [drawing functions](https://docs.godotengine.org/en/stable/classes/class_canvasitem.html#class-canvasitem-private-method-draw) ? ; draw_texture() is batched!, however drawing functions like draw_rect(), draw_circle(), draw_line(), draw_polygon(), etc... seem to make a separate draw call each time they are drawn (tested on 4.4 dev6). I think having them batched might attract people that look to game frameworks (like raylib, sdl2) for more performance with basic shapes..

**Comment by clayjohn** [comment]
> @stuartcarnie @clayjohn Awesome work!, any chance this batching could be extended to the other [drawing functions](https://docs.godotengine.org/en/stable/classes/class_canvasitem.html#class-canvasitem-private-method-draw) ? ; draw_texture(), draw_rect(), draw_line() are batched!, however drawing functions like draw_arc() draw_circle(), draw_polygon(), etc... seem to make a separate draw call each time they are drawn (tested on 4.4 dev6). I think having them batched might attract people that look to game frameworks (like raylib, sdl2) for more performance with basic shapes..

It's possible. We've discussed an approach to batching other primitives and I even wrote up some [notes on how it could work](https://github.com/orgs/godotengine/projects/33?pane=issue&itemId=20224421), but no one has been interested in implementing it. So feel free to take a stab at it if you want!

**Pull request merged**
